File: nodejs/src/sessionFsProvider.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import type {
    SessionFsHandler,
    SessionFsError,
    SessionFsStatResult,
    SessionFsReaddirWithTypesEntry,
} from "./generated/rpc.js";

/**
 * File metadata returned by {@link SessionFsProvider.stat}.
 * Same shape as the generated {@link SessionFsStatResult} but without the
 * `error` field, since providers signal errors by throwing.
 */
export type SessionFsFileInfo = Omit<SessionFsStatResult, "error">;

/**
 * Interface for session filesystem providers. Implementors use idiomatic
 * TypeScript patterns: throw on error, return values directly. Use
 * {@link createSessionFsAdapter} to convert a provider into the
 * {@link SessionFsHandler} expected by the SDK.
 *
 * Errors with a `code` property of `"ENOENT"` are mapped to the ENOENT
 * error code; all others map to UNKNOWN.
 */
export interface SessionFsProvider {
    /** Reads the full content of a file. Throw if the file does not exist. */
    readFile(path: string): Promise<string>;

    /** Writes content to a file, creating parent directories if needed. */
    writeFile(path: string, content: string, mode?: number): Promise<void>;

    /** Appends content to a file, creating parent directories if needed. */
    appendFile(path: string, content: string, mode?: number): Promise<void>;

    /** Checks whether a path exists. */
    exists(path: string): Promise<boolean>;

    /** Gets metadata about a file or directory. Throw if it does not exist. */
    stat(path: string): Promise<SessionFsFileInfo>;

    /** Creates a directory. If recursive is true, creates parents as needed. */
    mkdir(path: string, recursive: boolean, mode?: number): Promise<void>;

    /** Lists entry names in a directory. Throw if it does not exist. */
    readdir(path: string): Promise<string[]>;

    /** Lists entries with type info. Throw if the directory does not exist. */
    readdirWithTypes(path: string): Promise<SessionFsReaddirWithTypesEntry[]>;

    /** Removes a file or directory. If force is true, do not throw on ENOENT. */
    rm(path: string, recursive: boolean, force: boolean): Promise<void>;

    /** Renames/moves a file or directory. */
    rename(src: string, dest: string): Promise<void>;
}

/**
 * Wraps a {@link SessionFsProvider} into the {@link SessionFsHandler}
 * interface expected by the SDK, converting thrown errors into
 * {@link SessionFsError} results.
 */
export function createSessionFsAdapter(provider: SessionFsProvider): SessionFsHandler {
    return {
        readFile: async ({ path }) => {
            try {
                const content = await provider.readFile(path);
                return { content };
            } catch (err) {
                return { content: "", error: toSessionFsError(err) };
            }
        },
        writeFile: async ({ path, content, mode }) => {
            try {
                await provider.writeFile(path, content, mode);
                return undefined;
            } catch (err) {
                return toSessionFsError(err);
            }
        },
        appendFile: async ({ path, content, mode }) => {
            try {
                await provider.appendFile(path, content, mode);
                return undefined;
            } catch (err) {
                return toSessionFsError(err);
            }
        },
        exists: async ({ path }) => {
            try {
                return { exists: await provider.exists(path) };
            } catch {
                return { exists: false };
            }
        },
        stat: async ({ path }) => {
            try {
                return await provider.stat(path);
            } catch (err) {
                return {
                    isFile: false,
                    isDirectory: false,
                    size: 0,
                    mtime: new Date().toISOString(),
                    birthtime: new Date().toISOString(),
                    error: toSessionFsError(err),
                };
            }
        },
        mkdir: async ({ path, recursive, mode }) => {
            try {
                await provider.mkdir(path, recursive ?? false, mode);
                return undefined;
            } catch (err) {
                return toSessionFsError(err);
            }
        },
        readdir: async ({ path }) => {
            try {
                const entries = await provider.readdir(path);
                return { entries };
            } catch (err) {
                return { entries: [], error: toSessionFsError(err) };
            }
        },
        readdirWithTypes: async ({ path }) => {
            try {
                const entries = await provider.readdirWithTypes(path);
                return { entries };
            } catch (err) {
                return { entries: [], error: toSessionFsError(err) };
            }
        },
        rm: async ({ path, recursive, force }) => {
            try {
                await provider.rm(path, recursive ?? false, force ?? false);
                return undefined;
            } catch (err) {
                return toSessionFsError(err);
            }
        },
        rename: async ({ src, dest }) => {
            try {
                await provider.rename(src, dest);
                return undefined;
            } catch (err) {
                return toSessionFsError(err);
            }
        },
    };
}

function toSessionFsError(err: unknown): SessionFsError {
    const e = err as NodeJS.ErrnoException;
    const code = e.code === "ENOENT" ? "ENOENT" : "UNKNOWN";
    return { code, message: e.message ?? String(err) };
}

File: go/session_fs_provider.go
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package copilot

import (
	"errors"
	"os"
	"time"

	"github.com/github/copilot-sdk/go/rpc"
)

// SessionFSProvider is the interface that SDK users implement to provide
// a session filesystem. Methods use idiomatic Go error handling: return an
// error for failures (the adapter maps os.ErrNotExist → ENOENT automatically).
//
// To add SQLite support, also implement [SessionFSSqliteProvider] on the same type.
type SessionFSProvider interface {
	// ReadFile reads the full content of a file. Return os.ErrNotExist (or wrap it)
	// if the file does not exist.
	ReadFile(path string) (string, error)
	// WriteFile writes content to a file, creating it and parent directories if needed.
	// mode is an optional POSIX-style permission mode. Pass nil to use the OS default.
	WriteFile(path string, content string, mode *int) error
	// AppendFile appends content to a file, creating it and parent directories if needed.
	// mode is an optional POSIX-style permission mode. Pass nil to use the OS default.
	AppendFile(path string, content string, mode *int) error
	// Exists checks whether the given path exists.
	Exists(path string) (bool, error)
	// Stat returns metadata about a file or directory.
	// Return os.ErrNotExist if the path does not exist.
	Stat(path string) (*SessionFSFileInfo, error)
	// Mkdir creates a directory. If recursive is true, create parent directories as needed.
	// mode is an optional POSIX-style permission mode (e.g., 0o755). Pass nil to use the OS default.
	MakeDirectory(path string, recursive bool, mode *int) error
	// Readdir lists the names of entries in a directory.
	// Return os.ErrNotExist if the directory does not exist.
	ReadDirectory(path string) ([]string, error)
	// ReaddirWithTypes lists entries with type information.
	// Return os.ErrNotExist if the directory does not exist.
	ReadDirectoryWithTypes(path string) ([]rpc.SessionFSReaddirWithTypesEntry, error)
	// Rm removes a file or directory. If recursive is true, remove contents too.
	// If force is true, do not return an error when the path does not exist.
	Remove(path string, recursive bool, force bool) error
	// Rename moves/renames a file or directory.
	Rename(src string, dest string) error
}

// SessionFSSqliteProvider is an optional interface that a [SessionFSProvider]
// may also implement to support per-session SQLite databases. The adapter
// checks for this interface at runtime using a type assertion. If the
// provider does not implement it, SQLite requests return an "unsupported" error.
//
// Providers are already session-scoped (created per session by the factory),
// so these methods do not take a session ID parameter.
type SessionFSSqliteProvider interface {
	// SqliteQuery executes a SQLite query against the provider's per-session database.
	SqliteQuery(queryType rpc.SessionFSSqliteQueryType, query string, params map[string]any) (*SessionFSSqliteQueryResult, error)
	// SqliteExists checks whether the provider has a SQLite database for the session.
	SqliteExists() (bool, error)
}

// SessionFSSqliteQueryResult holds the result of a SQLite query execution.
// Same shape as the generated RPC type but without the Error field,
// since providers signal errors by returning a Go error.
type SessionFSSqliteQueryResult struct {
	Columns         []string         `json:"columns"`
	Rows            []map[string]any `json:"rows"`
	RowsAffected    int64            `json:"rowsAffected"`
	LastInsertRowid *int64           `json:"lastInsertRowid,omitempty"`
}

// SessionFSFileInfo holds file metadata returned by SessionFSProvider.Stat.
type SessionFSFileInfo struct {
	IsFile      bool
	IsDirectory bool
	Size        int64
	Mtime       time.Time
	Birthtime   time.Time
}

// sessionFSAdapter wraps a SessionFSProvider to implement rpc.SessionFSHandler,
// converting idiomatic Go errors into SessionFSError results.
type sessionFSAdapter struct {
	provider SessionFSProvider
}

func newSessionFSAdapter(provider SessionFSProvider) rpc.SessionFSHandler {
	return &sessionFSAdapter{provider: provider}
}

func (a *sessionFSAdapter) ReadFile(request *rpc.SessionFSReadFileRequest) (*rpc.SessionFSReadFileResult, error) {
	content, err := a.provider.ReadFile(request.Path)
	if err != nil {
		return &rpc.SessionFSReadFileResult{Error: toSessionFSError(err)}, nil
	}
	return &rpc.SessionFSReadFileResult{Content: content}, nil
}

func (a *sessionFSAdapter) WriteFile(request *rpc.SessionFSWriteFileRequest) (*rpc.SessionFSError, error) {
	var mode *int
	if request.Mode != nil {
		m := int(*request.Mode)
		mode = &m
	}
	if err := a.provider.WriteFile(request.Path, request.Content, mode); err != nil {
		return toSessionFSError(err), nil
	}
	return nil, nil
}

func (a *sessionFSAdapter) AppendFile(request *rpc.SessionFSAppendFileRequest) (*rpc.SessionFSError, error) {
	var mode *int
	if request.Mode != nil {
		m := int(*request.Mode)
		mode = &m
	}
	if err := a.provider.AppendFile(request.Path, request.Content, mode); err != nil {
		return toSessionFSError(err), nil
	}
	return nil, nil
}

func (a *sessionFSAdapter) Exists(request *rpc.SessionFSExistsRequest) (*rpc.SessionFSExistsResult, error) {
	exists, err := a.provider.Exists(request.Path)
	if err != nil {
		return &rpc.SessionFSExistsResult{Exists: false}, nil
	}
	return &rpc.SessionFSExistsResult{Exists: exists}, nil
}

func (a *sessionFSAdapter) Stat(request *rpc.SessionFSStatRequest) (*rpc.SessionFSStatResult, error) {
	info, err := a.provider.Stat(request.Path)
	if err != nil {
		return &rpc.SessionFSStatResult{Error: toSessionFSError(err)}, nil
	}
	return &rpc.SessionFSStatResult{
		IsFile:      info.IsFile,
		IsDirectory: info.IsDirectory,
		Size:        info.Size,
		Mtime:       info.Mtime,
		Birthtime:   info.Birthtime,
	}, nil
}

func (a *sessionFSAdapter) Mkdir(request *rpc.SessionFSMkdirRequest) (*rpc.SessionFSError, error) {
	recursive := request.Recursive != nil && *request.Recursive
	var mode *int
	if request.Mode != nil {
		m := int(*request.Mode)
		mode = &m
	}
	if err := a.provider.MakeDirectory(request.Path, recursive, mode); err != nil {
		return toSessionFSError(err), nil
	}
	return nil, nil
}

func (a *sessionFSAdapter) Readdir(request *rpc.SessionFSReaddirRequest) (*rpc.SessionFSReaddirResult, error) {
	entries, err := a.provider.ReadDirectory(request.Path)
	if err != nil {
		return &rpc.SessionFSReaddirResult{Error: toSessionFSError(err)}, nil
	}
	return &rpc.SessionFSReaddirResult{Entries: entries}, nil
}

func (a *sessionFSAdapter) ReaddirWithTypes(request *rpc.SessionFSReaddirWithTypesRequest) (*rpc.SessionFSReaddirWithTypesResult, error) {
	entries, err := a.provider.ReadDirectoryWithTypes(request.Path)
	if err != nil {
		return &rpc.SessionFSReaddirWithTypesResult{Error: toSessionFSError(err)}, nil
	}
	return &rpc.SessionFSReaddirWithTypesResult{Entries: entries}, nil
}

func (a *sessionFSAdapter) Rm(request *rpc.SessionFSRmRequest) (*rpc.SessionFSError, error) {
	recursive := request.Recursive != nil && *request.Recursive
	force := request.Force != nil && *request.Force
	if err := a.provider.Remove(request.Path, recursive, force); err != nil {
		return toSessionFSError(err), nil
	}
	return nil, nil
}

func (a *sessionFSAdapter) Rename(request *rpc.SessionFSRenameRequest) (*rpc.SessionFSError, error) {
	if err := a.provider.Rename(request.Src, request.Dest); err != nil {
		return toSessionFSError(err), nil
	}
	return nil, nil
}

func (a *sessionFSAdapter) SqliteQuery(request *rpc.SessionFSSqliteQueryRequest) (*rpc.SessionFSSqliteQueryResult, error) {
	sp, ok := a.provider.(SessionFSSqliteProvider)
	if !ok {
		msg := "SQLite is not supported by this session filesystem provider"
		return &rpc.SessionFSSqliteQueryResult{
			Columns:      []string{},
			Rows:         []map[string]any{},
			RowsAffected: 0,
			Error:        &rpc.SessionFSError{Code: rpc.SessionFSErrorCodeUNKNOWN, Message: &msg},
		}, nil
	}
	result, err := sp.SqliteQuery(request.QueryType, request.Query, request.Params)
	if err != nil {
		return &rpc.SessionFSSqliteQueryResult{
			Columns:      []string{},
			Rows:         []map[string]any{},
			RowsAffected: 0,
			Error:        toSessionFSError(err),
		}, nil
	}
	if result == nil {
		return &rpc.SessionFSSqliteQueryResult{
			Columns:      []string{},
			Rows:         []map[string]any{},
			RowsAffected: 0,
		}, nil
	}
	var wireRowid *int64
	if result.LastInsertRowid != nil {
		rowid := *result.LastInsertRowid
		wireRowid = &rowid
	}
	return &rpc.SessionFSSqliteQueryResult{
		Columns:         result.Columns,
		Rows:            result.Rows,
		RowsAffected:    result.RowsAffected,
		LastInsertRowid: wireRowid,
	}, nil
}

func (a *sessionFSAdapter) SqliteExists(request *rpc.SessionFSSqliteExistsRequest) (*rpc.SessionFSSqliteExistsResult, error) {
	sp, ok := a.provider.(SessionFSSqliteProvider)
	if !ok {
		return &rpc.SessionFSSqliteExistsResult{Exists: false}, nil
	}
	exists, err := sp.SqliteExists()
	if err != nil {
		return &rpc.SessionFSSqliteExistsResult{Exists: false}, nil
	}
	return &rpc.SessionFSSqliteExistsResult{Exists: exists}, nil
}

func toSessionFSError(err error) *rpc.SessionFSError {
	code := rpc.SessionFSErrorCodeUNKNOWN
	if errors.Is(err, os.ErrNotExist) {
		code = rpc.SessionFSErrorCodeENOENT
	}
	msg := err.Error()
	return &rpc.SessionFSError{Code: code, Message: &msg}
}
